src/dynamics/minimal.lean

```diff
@@ -0,0 +1,105 @@
+/-
+Copyright (c) 2021 Yury G. Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury G. Kudryashov
+-/
+import topology.algebra.mul_action
+import group_theory.group_action.basic
+
+/-!
+# Minimal action of a group
+
+In this file we define an action of a monoid `M` on a topological space `α` to be *minimal* if the
+`M`-orbit of every point `x : α` is dense. We also provide an additive version of this definiton and
+prove some basic facts about minimal actions.
+
+## TODO
+
+* Define a minimal set of an action.
+
+## Tags
+
+group action, minimal
+-/
+
+open_locale pointwise
+open set
+
+/-- An action of an additive monoid `M` on a topological space is called *minimal* if the `M`-orbit
+of every point `x : α` is dense. -/
+class add_action.is_minimal (M α : Type*) [add_monoid M] [topological_space α] [add_action M α] :
+  Prop :=
+(dense_orbit : ∀ x : α, dense (add_action.orbit M x))
+
+/-- An action of a monoid `M` on a topological space is called *minimal* if the `M`-orbit of every
+point `x : α` is dense. -/
+@[to_additive]
+class mul_action.is_minimal (M α : Type*) [monoid M] [topological_space α] [mul_action M α] :
+  Prop :=
+(dense_orbit : ∀ x : α, dense (mul_action.orbit M x))
+
+namespace mul_action
+
+variables (M G : Type*) {α : Type*} [monoid M] [group G] [topological_space α] [mul_action M α]
+  [mul_action G α]
+
+@[to_additive] lemma dense_orbit [is_minimal M α] (x : α) : dense (orbit M x) :=
+mul_action.is_minimal.dense_orbit x
+
+@[to_additive] lemma dense_range_smul [is_minimal M α] (x : α) :
+  dense_range (λ c : M, c • x) :=
+dense_orbit M x
+
+@[priority 100, to_additive]
+instance is_minimal_of_pretransitive [is_pretransitive M α] : is_minimal M α :=
+⟨λ x, (surjective_smul M x).dense_range⟩
+
+@[to_additive]
+lemma exists_smul_mem_open [is_minimal M α] (x : α) {U : set α} (hUo : is_open U)
+  (hne : U.nonempty) : ∃ c : M, c • x ∈ U :=
+(dense_range_smul M x).exists_mem_open hUo hne
+
+@[to_additive]
+lemma exists_cover_smul [is_minimal G α] (s : set α) {U : set α} (hUo : is_open U)
+  (hne : U.nonempty) :
+  ∃ I : set G, s ⊆ ⋃ g ∈ I, g • U :=
```

Comment:
this one is strange: the conclusion is true by taking `I = univ`, so why bother stating it as a `bUnion` instead of just the `Union`? and then why not just state that the union of `g • U` is equal to `univ`?

Reply:
Thanks! Fixed. I also renamed most lemmas to avoid `mul_action` namespace when the name contains `smul`.